stringdtype/stringdtype/src/dtype.c

```diff
@@ -109,58 +132,74 @@ stringdtype_setitem(StringDTypeObject *NPY_UNUSED(descr), PyObject *obj,
                     char **dataptr)
 {
     PyObject *val_obj = get_value(obj);
+
     if (val_obj == NULL) {
         return -1;
     }
 
-    char *val = NULL;
-    Py_ssize_t length = 0;
-    if (PyBytes_AsStringAndSize(val_obj, &val, &length) == -1) {
-        return -1;
-    }
+    ss *sdata = (ss *)dataptr;
 
     // free if dataptr holds preexisting string data,
     // ssfree does a NULL check
-    ssfree((ss *)dataptr);
+    ssfree(sdata);
 
-    // copies contents of val into item_val->buf
-    int res = ssnewlen(val, length, (ss *)dataptr);
+    // RichCompareBool short-circuits to a pointer comparison fast-path
+    // so no need to do pointer comparison first
+    int eq_res = PyObject_RichCompareBool(val_obj, NA_OBJ, Py_EQ);
 
-    // val_obj must stay alive until here to ensure *val* doesn't get
-    // deallocated
-    Py_DECREF(val_obj);
+    if (eq_res < 0) {
+        goto error;
+    }
 
-    if (res == -1) {
-        PyErr_NoMemory();
-        return -1;
+    if (eq_res == 1) {
+        sdata = NULL;
     }
-    else if (res == -2) {
-        // this should never happen
-        assert(0);
+    else {
+        char *val = NULL;
+        Py_ssize_t length = 0;
+        if (PyBytes_AsStringAndSize(val_obj, &val, &length) == -1) {
+            goto error;
+        }
+
+        // copies contents of val into item_val->buf
+        int res = ssnewlen(val, length, sdata);
+
+        if (res == -1) {
+            PyErr_NoMemory();
+            goto error;
+        }
+        else if (res == -2) {
+            // this should never happen
+            assert(0);
+            goto error;
+        }
     }
 
+    Py_DECREF(val_obj);
     return 0;
+
+error:
+    Py_DECREF(val_obj);
+    return -1;
 }
 
 static PyObject *
 stringdtype_getitem(StringDTypeObject *NPY_UNUSED(descr), char **dataptr)
 {
-    char *data;
-    size_t len;
+    PyObject *val_obj = NULL;
+    ss *sdata = (ss *)dataptr;
 
-    if (*dataptr == NULL) {
-        data = "\0";
-        len = 0;
+    if (ss_isnull(sdata)) {
+        Py_INCREF(NA_OBJ);
+        val_obj = NA_OBJ;
     }
     else {
-        data = ((ss *)dataptr)->buf;
-        len = ((ss *)dataptr)->len;
-    }
-
-    PyObject *val_obj = PyUnicode_FromStringAndSize(data, len);
-
-    if (val_obj == NULL) {
-        return NULL;
+        char *data = sdata->buf;
+        size_t len = sdata->len;
+        val_obj = PyUnicode_FromStringAndSize(data, len);
+        if (val_obj == NULL) {
+            return NULL;
+        }
```

Comment:
I think this check is unnecessary: `val_obj` gets returned in either case.

Reply:
I'd prefer to keep this just in case someone adds any further logic after this that might use `val_obj`.